pkg/proxy/proxy.go

```diff
@@ -18,43 +18,78 @@ import (
 
 	"github.com/teslamotors/vehicle-command/internal/authentication"
 	"github.com/teslamotors/vehicle-command/internal/log"
-	"github.com/teslamotors/vehicle-command/pkg/account"
 	"github.com/teslamotors/vehicle-command/pkg/cache"
 	"github.com/teslamotors/vehicle-command/pkg/connector/inet"
 	"github.com/teslamotors/vehicle-command/pkg/protocol"
-	"github.com/teslamotors/vehicle-command/pkg/vehicle"
+	"github.com/teslamotors/vehicle-command/pkg/protocol/protobuf/universalmessage"
 )
 
 const (
 	DefaultTimeout       = 10 * time.Second
-	maxRequestBodyBytes  = 512
 	vinLength            = 17
 	proxyProtocolVersion = "tesla-http-proxy/1.1.0"
+	xForwardedForHeader  = "X-Forwarded-For"
 )
 
-func getAccount(req *http.Request) (*account.Account, error) {
-	token, ok := strings.CutPrefix(req.Header.Get("Authorization"), "Bearer ")
-	if !ok {
-		return nil, fmt.Errorf("client did not provide an OAuth token")
-	}
-	return account.New(token, proxyProtocolVersion)
+type Vehicle interface {
+	Connect(context.Context) error
+	Disconnect()
+	StartSession(context.Context, []universalmessage.Domain) error
+	UpdateCachedSessions(*cache.SessionCache) error
+	ExecuteAction(context.Context, interface{}) error
+}
+
+type Account interface {
+	GetVehicle(context.Context, string, authentication.ECDHPrivateKey, *cache.SessionCache) (Vehicle, error)
+	GetHost() string
+}
+
+type AccountProvider func(oauthToken, userAgent string) (Account, error)
+
+// Response contains a server's response to a client request.
+type Response struct {
+	Response   interface{} `json:"response"`
+	Error      string      `json:"error"`
+	ErrDetails string      `json:"error_description"`
+}
+
+// vehicleResponse is the response format used by the vehicle's command API.
+type vehicleResponse struct {
+	Result bool   `json:"result"`
+	Reason string `json:"string"`
+}
+
+// connectionHeaders are HTTP headers that are not forwarded.
+var connectionHeaders = []string{
+	"Proxy-Connection",
+	"Keep-Alive",
+	"Transfer-Encoding",
+	"Te",
+	"Upgrade",
 }
 
+// contextKey is a type used to store values in context.Context.
+type contextKey string
+
+const accountContext contextKey = "account"
+
 // Proxy exposes an HTTP API for sending vehicle commands.
 type Proxy struct {
 	Timeout time.Duration
 
-	commandKey  protocol.ECDHPrivateKey
-	sessions    *cache.SessionCache
-	vinLock     sync.Map
-	unsupported sync.Map
+	commandKey      protocol.ECDHPrivateKey
+	sessions        *cache.SessionCache
+	vinLock         sync.Map
+	unsupported     sync.Map
+	handler         http.Handler
+	accountProvider AccountProvider
 }
 
-func (p *Proxy) markUnsupportedVIN(vin string) {
+func (p *Proxy) markSignedCommandsUnsupportedVIN(vin string) {
 	p.unsupported.Store(vin, true)
 }
 
-func (p *Proxy) isNotSupported(vin string) bool {
+func (p *Proxy) signedCommandUnsupported(vin string) bool {
```

Comment:
Previous one says `Commands`, this one says `Command`, but they're dealing with the same thing.